File: sdk/typescript/src/hashchain.ts
// SPDX-FileCopyrightText: 2026 The Attestplane Authors
// SPDX-License-Identifier: Apache-2.0
/**
 * Pure-function hash-chain primitives for the Attestplane substrate.
 *
 * Mirrors `sdk/python/src/attestplane/hashchain.py`. Pure: no I/O, no global
 * state, no time reads. Container responsibility lives in `substrate.ts`.
 */

import { createHash } from 'node:crypto';
import { v7 as uuidv7 } from 'uuid';

import { canonicalize } from './canonical.js';
import type { AuditEvent, ChainHead, ChainedEvent, EventDraft } from './types.js';

export const SCHEMA_VERSION = 1;
export const GENESIS_HASH: Uint8Array = new Uint8Array(32);

export interface VerificationResult {
  readonly ok: boolean;
  readonly first_bad_index: number | null;
  readonly reason: string | null;
}

export function genesisHead(): ChainHead {
  return { seq: -1, event_hash: GENESIS_HASH };
}

export function hashEvent(event: AuditEvent): Uint8Array {
  return new Uint8Array(createHash('sha256').update(canonicalizeAuditEvent(event)).digest());
}

export function canonicalizeAuditEvent(event: AuditEvent): Uint8Array {
  return canonicalize(canonicalEvent(event));
}

export interface ChainExtendOptions {
  readonly now: Date;
  readonly event_id?: string;
}

export function chainExtend(
  tip: ChainHead,
  draft: EventDraft,
  options: ChainExtendOptions,
): ChainedEvent {
  if (Number.isNaN(options.now.getTime())) {
    throw new Error('chainExtend requires a valid Date for now');
  }
  const event_id = options.event_id ?? uuidv7();
  const event: AuditEvent = {
    schema_version: SCHEMA_VERSION,
    event_id,
    timestamp: options.now,
    event_type: draft.event_type,
    actor: draft.actor,
    payload: draft.payload,
    subject_ref: draft.subject_ref,
    session_id: draft.session_id,
    reference_db_ref: draft.reference_db_ref,
    matched_input_ref: draft.matched_input_ref,
    human_verifier: draft.human_verifier,
  };
  const event_hash = hashEvent(event);
  return {
    seq: tip.seq + 1,
    prev_hash: tip.event_hash,
    event_hash,
    event,
  };
}

export function verifyChain(events: readonly ChainedEvent[]): VerificationResult {
  let expectedTip = genesisHead();
  for (let i = 0; i < events.length; i++) {
    const item = events[i] as ChainedEvent;
    if (item.seq !== i) {
      return {
        ok: false,
        first_bad_index: i,
        reason: `seq mismatch at index ${i}: got ${item.seq}, expected ${i}`,
      };
    }
    if (!bytesEqual(item.prev_hash, expectedTip.event_hash)) {
      return {
        ok: false,
        first_bad_index: i,
        reason: `prev_hash mismatch at seq ${i}`,
      };
    }
    const recomputed = hashEvent(item.event);
    if (!bytesEqual(recomputed, item.event_hash)) {
      return {
        ok: false,
        first_bad_index: i,
        reason: `event_hash mismatch at seq ${i}`,
      };
    }
    expectedTip = { seq: item.seq, event_hash: item.event_hash };
  }
  return { ok: true, first_bad_index: null, reason: null };
}

export function headOf(events: readonly ChainedEvent[]): ChainHead {
  if (events.length === 0) return genesisHead();
  const last = events[events.length - 1] as ChainedEvent;
  return { seq: last.seq, event_hash: last.event_hash };
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function canonicalEvent(event: AuditEvent): Record<string, unknown> {
  const timestamp = (event as { readonly timestamp: unknown }).timestamp;
  return {
    ...event,
    timestamp:
      timestamp instanceof Date ? formatDateUtcMicroseconds(timestamp, '$.timestamp') : timestamp,
  };
}

function formatDateUtcMicroseconds(value: Date, path: string): string {
  const ms = value.getTime();
  if (Number.isNaN(ms)) {
    throw new Error(`${path}: invalid Date (NaN time value)`);
  }
  const yyyy = value.getUTCFullYear().toString().padStart(4, '0');
  const mm = (value.getUTCMonth() + 1).toString().padStart(2, '0');
  const dd = value.getUTCDate().toString().padStart(2, '0');
  const hh = value.getUTCHours().toString().padStart(2, '0');
  const mi = value.getUTCMinutes().toString().padStart(2, '0');
  const ss = value.getUTCSeconds().toString().padStart(2, '0');
  const milli = value.getUTCMilliseconds().toString().padStart(3, '0');
  return `${yyyy}-${mm}-${dd}T${hh}:${mi}:${ss}.${milli}000Z`;
}

File: sdk/python/src/attestplane/hashchain.py
# SPDX-FileCopyrightText: 2026 The Attestplane Authors
# SPDX-License-Identifier: Apache-2.0
"""Pure-function hash-chain primitives.

This module is intentionally I/O-free and stateless. All state lives in the
caller (``AttestSubstrate`` or a future multi-writer backend). Every function
here is referentially transparent, which makes the substrate semantics
auditable independently of any particular storage implementation.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Final

from attestplane.canonical import canonicalize
from attestplane.types import AuditEvent, ChainedEvent, ChainHead, EventDraft

SCHEMA_VERSION: Final[int] = 1
"""Canonicalization schema version for v0.0.1. Frozen by conformance vectors."""

GENESIS_HASH: Final[bytes] = b"\x00" * 32
"""``prev_hash`` value for the first event in any chain."""


@dataclass(frozen=True, slots=True)
class VerificationResult:
    """Outcome of walking and re-hashing a chain.

    ``first_bad_index`` is the seq of the first event that failed validation,
    or ``None`` if the chain is consistent. ``reason`` describes the failure
    in human-readable form.
    """

    ok: bool
    first_bad_index: int | None
    reason: str | None


def genesis_head() -> ChainHead:
    """Return the head used before any event has been appended."""
    return ChainHead(seq=-1, event_hash=GENESIS_HASH)


def hash_event(event: AuditEvent) -> bytes:
    """Compute the SHA-256 digest of an ``AuditEvent``'s canonical encoding."""
    return hashlib.sha256(canonicalize(event)).digest()


def chain_extend(
    tip: ChainHead,
    draft: EventDraft,
    *,
    now: datetime,
    event_id: str | None = None,
) -> ChainedEvent:
    """Extend the chain by one event.

    Pure function: given the same tip, draft, and clock, this produces a
    deterministic ``ChainedEvent`` (assuming ``event_id`` is provided). If
    ``event_id`` is ``None``, a UUIDv7 is generated from ``now``, which
    introduces non-determinism through the random component but preserves the
    time ordering needed for index locality.

    The substrate enforces append-order; this function does not verify that
    ``tip`` is the actual current tip — that responsibility lives in the
    container (``AttestSubstrate.append`` or its multi-writer successor).
    """
    if now.tzinfo is None or now.utcoffset() != UTC.utcoffset(None):
        raise ValueError("chain_extend requires a UTC-aware datetime for 'now'")

    if event_id is not None:
        resolved_event_id = event_id
    else:
        import uuid_utils

        resolved_event_id = str(uuid_utils.uuid7())
    event = AuditEvent(
        schema_version=SCHEMA_VERSION,
        event_id=resolved_event_id,
        timestamp=now,
        event_type=draft.event_type,
        actor=draft.actor,
        payload=draft.payload,
        subject_ref=draft.subject_ref,
        session_id=draft.session_id,
        reference_db_ref=draft.reference_db_ref,
        matched_input_ref=draft.matched_input_ref,
        human_verifier=draft.human_verifier,
    )
    event_hash = hash_event(event)
    return ChainedEvent(
        seq=tip.seq + 1,
        prev_hash=tip.event_hash,
        event_hash=event_hash,
        event=event,
    )


def verify_chain(events: list[ChainedEvent]) -> VerificationResult:
    """Walk a chain in order, re-hashing each event and checking linkage.

    Returns ``ok=True`` for an empty chain (a substrate with no appended
    events). Returns the seq of the first event that fails validation
    otherwise.
    """
    expected_tip = genesis_head()
    for index, item in enumerate(events):
        if item.seq != index:
            return VerificationResult(
                ok=False,
                first_bad_index=index,
                reason=f"seq mismatch at index {index}: got {item.seq}, expected {index}",
            )
        if item.prev_hash != expected_tip.event_hash:
            return VerificationResult(
                ok=False,
                first_bad_index=index,
                reason=f"prev_hash mismatch at seq {index}",
            )
        recomputed = hash_event(item.event)
        if recomputed != item.event_hash:
            return VerificationResult(
                ok=False,
                first_bad_index=index,
                reason=f"event_hash mismatch at seq {index}",
            )
        expected_tip = ChainHead(seq=item.seq, event_hash=item.event_hash)
    return VerificationResult(ok=True, first_bad_index=None, reason=None)


def head_of(events: list[ChainedEvent]) -> ChainHead:
    """Return the current head of a chain (``genesis_head()`` when empty)."""
    if not events:
        return genesis_head()
    last = events[-1]
    return ChainHead(seq=last.seq, event_hash=last.event_hash)


__all__ = [
    "GENESIS_HASH",
    "SCHEMA_VERSION",
    "VerificationResult",
    "chain_extend",
    "genesis_head",
    "hash_event",
    "head_of",
    "verify_chain",
]
